deploy/runtools/run_farm.py

```diff
@@ -804,3 +810,500 @@ def terminate_by_inst(self, inst: Inst) -> None:
+            local_tmp_path = local_db.name
+            
+        rootLogger.info(f"Reading database file {db_path} from host {host_machine_ip} to local temp file {local_tmp_path}")
+        
+        # Acquire lock, copy file
+        try:
+            subprocess.run([
+                "ssh", host_machine_ip,
+                f"flock {lock_path} cat {db_path}"
+            ], check=True, stdout=open(local_tmp_path, "wb"))
+        except subprocess.CalledProcessError as e:
+            raise RuntimeError(f"Failed to read database file {db_path} on host {host_machine_ip}: {e} -- Did you run sudo ./setup-firesim-vm-host?")
+    
+        # Step 2: Load + modify JSON locally
+        with open(local_tmp_path, "r") as db_file:
+            try:
+                host_machine_fpgas_db = json.load(db_file)
+            except json.JSONDecodeError as e:
+                raise ValueError(f"Error decoding JSON from {db_path}: {e}")
+        rootLogger.info(f"Loaded database from {db_path}, locally.")
+
+        # get number of FPGAs that are not busy
+        free_fpgas = [
+            fpga for fpga in host_machine_fpgas_db if not host_machine_fpgas_db[fpga]["busy"]
+        ]
+        rootLogger.info(f"Number of free FPGAs: {len(free_fpgas)}")
+
+        # get number of FPGAs requested by the spec
+        num_requested_fpgas = self.SIM_HOST_HANDLE_TO_MAX_FPGA_SLOTS[host_machine_ip]
+
+        rootLogger.info(f"Number of FPGAs requested by the spec: {num_requested_fpgas}")
+
+        # check if there are enough free FPGAs available
+        if len(free_fpgas) < num_requested_fpgas:
+            raise RuntimeError(
+                f"Not enough free FPGAs available. Requested: {num_requested_fpgas}, Available: {len(free_fpgas)}"
+            )
+
+        # grab their bdfs and attach them to the VM, set them as busy, record vm IP in the database
+        
+        fpgas_to_attach = free_fpgas[:num_requested_fpgas]
+        
+        # set those fpgas as busy in the database & write to file
+        for fpga in fpgas_to_attach:
+            host_machine_fpgas_db[fpga]["busy"] = True
+            host_machine_fpgas_db[fpga]["vm_name"] = self.vm_name  # use the VM name + username to avoid collisions
+            host_machine_fpgas_db[fpga]["in_use_by"] = local("whoami", capture=True)
+        
+        # Step 3: Write it back to file
+        with open(local_tmp_path, "w") as db_file:
+            try:
+                json.dump(host_machine_fpgas_db, db_file, indent=2)
+            except TypeError as e:
+                raise ValueError(f"Error writing JSON to {db_path}: {e}")
+        rootLogger.info(f"Updated database with attached FPGAs: {json.dumps(host_machine_fpgas_db, indent=2)}")
+        
+        # Step 4: scp back while holding the lock
+        subprocess.run([
+            "ssh", host_machine_ip,
+            f"flock {lock_path} bash -c 'cat > {db_path}'"
+        ], input=open(local_tmp_path, "rb").read(), check=True)
```

Comment:
Why not use Fabric instead of directly calling subprocess with ssh? You should be able to write a helper function + `execute` to do the same thing: https://github.com/firesim/firesim/blob/e8c384a26737d667c909c0289e44564a18c54854/deploy/firesim#L310